### DISCUSSION TITLE
How to show command options only with preceding "-"?

### DISCUSSION THREAD

**Discussion by moovinzoo** [akinomyoga/ble.sh#436] [Q&A] [answered]
For now, when I use command ~~"vim [tab]"~~ `git <tab>` then auto-complete menu shows files(candidates) and vim command's options.
I guess the options are sourced by man.
But 1. I want inner command to be shown, 2. I don't want command options to be shown without '-'.

I read the discussion here briefly and guess [this commit](https://github.com/akinomyoga/ble.sh/commit/54ace59c57557169ce68366f23849cfe5f8413e8#diff-c180d07f9f6314cd0ce60494fd95e4a1a3afaa2f78afd7046897c4c4e6f072d3R5122-R5141) is about my requirement.
But I don't understand how to configure and even cannot sure that code fragment is about this topic.

To clarify, I want the command options only when - is preceding after certain command.
For example,
(1) ~~vim [tab]~~ `git <tab>` shows only (sub)commands and candidates(files)
(2) ~~vim -[tab]~~ `git -<tab>` shows only command options

May be super helpful if you could share your knowledge.
Thanks a lot!

FYI. I'm using `man-db` for man command.

**Comment by moovinzoo** [comment]
I think this behavior has been changed partially by the timeline below.
My system(arch) originally doesn't have man, I installed vim, git, bash, ble.sh first.
After I installed man and tried configuring blesh with its completion functionality, After I installed bash-completion and found that git is not in the completion list. Now I realize it is so weird. So I tried to reinstall pre-installed packages then bash-completion/completions now listed the completion about that packages. And shows  (sub)commands in menu by pressing tab after certain command.

But it is not that consistent. When I use command vim (tab) or git (tab), it shows options, files for the first time when I enters new shell. But when I delete the space before command and re-insert space and press tab then it shows (sub)commands as expected.
So hard to narrow down the reason to me..

**Reply by akinomyoga** [reply]
> When I use command vim (tab) or git (tab), it shows options, files for the first time when I enters new shell. But when I delete the space before command and re-insert space and press tab then it shows (sub)commands as expected.

Thanks. It doesn't seem to reproduce in my environment. With `bash-completion`, it already behaves in the latter way from the beginning in my environment. Without `bash-completion`, it always shows both the files and options as in the former way (which is expected because that is the behavior of built-in completion generation of ble.sh), and the behavior doesn't change after deleting and inserting spaces.

For example, do you set up bash-completion in a special way? If you could share the content of your `~/.bashrc`, it might help.

**Reply by moovinzoo** [reply]
Sure! ~/.bashrc:
```
#
# ~/.bashrc
#

# If not running interactively, don't do anything
[[ $- != *i* ]] && return

# Enable ble.sh (1/2)
source ~/.local/share/blesh/ble.sh --noattach --rcfile "$HOME/.blerc"

# Normal settings can be placed in here.

PS1='[\u@\h \W]\$ '
HISTSIZE=1000
HISTFILESIZE=2000

# Enable fzf
eval "$(fzf --bash)"

# Enable zoxide
eval "$(zoxide init bash)"

# # Enable starship
# eval "$(starship init bash)"

# Enable ble.sh (2/2)
[[ ${BLE_VERSION-} ]] && ble-attach

**Reply by moovinzoo** [reply]
~/.blerc:
```
# blerc

# input encoding
bleopt input_encoding=UTF-8

# use bash-completion, if available
[[ $PS1 && -f /usr/share/bash-completion/bash_completion ]] && \
    . /usr/share/bash-completion/bash_completion

# fzf integration
_ble_contrib_fzf_base=/usr/share/fzf
# ble-import integration/fzf-completion
ble-import integration/fzf-key-bindings

```

**Reply by moovinzoo** [reply]
Lastly, ~/.bash_profile,
```
#
# ~/.bash_profile
#

[[ -f ~/.bashrc ]] && . ~/.bashrc
```

**Reply by akinomyoga** [reply]
Thanks. Can you check the behavior when you temporarily turn off Fzf and Zoxide settings by commenting out the corresponding lines?

Could you also check the result of the following command in a ble.sh session?

```bash
$ ble/widget/display-shell-version
```

**Reply by moovinzoo** [reply]
If I disabled both fzf, zioxide then inner-commands shows up from the first time.
  I've already commented the fzf-completion functionality to clear things up, but there should be confliction in else-where, as you suggested.
```
$ ble/widget/display-shell-version

GNU bash, version 5.2.26(1)-release (x86_64-pc-linux-gnu) [Arch Linux]
ble.sh, version 0.4.0-devel4+adf9ab8 (noarch) [git 2.44.0, GNU Make 4.4.1, GNU Awk 5.3.0, API 4.0, PMA Avon 8-g1, (GNU MPFR 4.2.1, GNU MP 6.3.0)]
bash-completion, version 2.11 (hash:2d02f73e803daa87a06e94e33b2a7b3e672a2c0c, 76338 by tes) (noarch)
locale: LANG=es_US.UTF-8
terminal: TERM=xterm-kitty wcwidth=15.0-west/15.1-2+ri, kitty:0 (1;4000;33)
```

**Reply by moovinzoo** [reply]
Additionally, even if I enable zioxide again, completion has no problem anymore.
But If I enable FZF alone, weird behavior starts again.
It seems pretty clear that FZF was the reason.
I don't have any configuration on FZF, it is on the default configuration, installed via pacman.
(FYI)
```
$ fzf --version

0.48.1 (9ffe951f)
```

**Reply by moovinzoo** [reply]
![weird_completion_for_the_first_time](https://github.com/akinomyoga/ble.sh/assets/28827932/76ebcae0-b728-46f9-8854-7db48f7e1232)
![completion_as_expected_when_retab](https://github.com/akinomyoga/ble.sh/assets/28827932/0efbe1d3-6db0-43f9-9a2d-f9e886ca9909)

Sorry about the large size of screenshot, cause I'm not familiar with current system

**Reply by akinomyoga** [reply]
Thank you for those pieces of information. I've tried your settings of `~/.bashrc` and `~/.blerc` (with Bash 5.2.21, bash-completion 2.12, Fzf 0.48.0 (f97d2754), and Zoxide 0.9.2), but I cannot still reproduce the problem. The subcommands always show up from the beginning in my environment just the same as in your second picture.

**Reply by moovinzoo** [reply]
Oh, thank you very much for your effort.
I reinstalled fzf using package manager but it does not resolve the situation.
Actually I installed ble.sh from git with using given installing script instead of using Arch package manager(pacman).
That was because I found the difference of behavior between two along the setup process, originally I installed blesh with pacman but I cannot remember what it was exactly.
But from this point, is there any possibility that fzf's default configuration could be different on what installing process was?

**Reply by akinomyoga** [reply]
OK, I could reproduce it in Arch Linux. I was testing it in Fedora and cannot still reproduce it in Fedora with exactly the same bashrc/blerc, but I can test it in my Arch Linux system.

**Reply by akinomyoga** [reply]
> originally I installed blesh with pacman but I cannot remember what it was exactly.

Is it probably an AUR helper instead of `pacman` (because `blesh` is not provided in the official Arch package repository)?

In AUR, there are two versions of ble.sh, ble-0.3 (provided with the name [`blesh`](https://aur.archlinux.org/packages/blesh)) and ble-0.4 (provided with the name [`blesh-git`](https://aur.archlinux.org/packages/blesh-git)). ble-0.3 is significantly old and the compatibility with other frameworks are not good, so I anyway recommend you to use the latest version of ble-0.4 either manually installed or installed by an AUR helper.

**Reply by moovinzoo** [reply]
Oh, sorry. Yes it was by AUR helper, `paru`.
I prefer the lts-like version so maybe it was (what I felt feature is quite different from the description) `blesh`, the old 0.3 one.

**Reply by moovinzoo** [reply]
OK, as you recommended, I'll keep using manually installed current one.

**Reply by moovinzoo** [reply]
I temporarily settled down on pretty weird state that not enabling fzf on .bashrc but enable ops in .blerc.
Cannot use ** wildcard in command but all functionality including FZF command like C-R, C-T, A-C are working.
Not showing default menu completion from the beginning gives me comfort.

**Reply by akinomyoga** [reply, marked as answer]
You seem to have hit a bug of Fzf in the specific version you use, 0.48.1. First, I could reproduce the problem without ble.sh with the following bashrc:

```bash
source /usr/share/bash-completion/bash_completion
eval "$(fzf --bash)"
bind 'set show-all-if-ambiguous on'
```

Then, I checked the difference in the debug trace produced by `set -x` between Fedora and Arch. It turned out the completion scripts are different. Then, I checked the different versions of fzf. The commit https://github.com/junegunn/fzf/commit/0740ef7ceb717676c7d8328a655e1554e1b0d366 (deployed in 0.48.1) seems to have introduced the bug, and the commit https://github.com/junegunn/fzf/commit/dff865239a0cdc79f33e97ec900b2bc8dd6517d2 (deployed in 0.49.0) seems to have fixed the problem. This is the report upstream: https://github.com/junegunn/fzf/issues/3702.

**Reply by moovinzoo** [reply]
Sorry for the late replying. 
```
$ paru -Qs fzf
local/fzf 0.48.1-1
    Command-line fuzzy finder
```
As you mensioned, I installed fzf using paru(AUR helper), and it shows that the exact old version that you said.
I misunderstood that if a package is listed in package manager, it also guarantees that latest(at least close stable) version.
Glad to hear that fzf now solved that integration problem with yours.
Besides sharing the process you narrowing down the reason, was also helpful for me to understand how packages integrates with each other.
Thank you for your great plugin also, I'm enjoying it!

**Reply by akinomyoga** [reply]
Thanks!

**Comment by akinomyoga** [comment]
Providing the detailed behavior of the completion settings is not a goal of ble.sh. Instead, ble.sh lets external completion settings to control the behavior.

> I guess the options are sourced by man.

Yes. Or, more specifically, the *descriptions* of the options are retrieved from the man pages. The generation of options for completion depends on the setup. With the plain setup without `bash-completion` or any other completion settings, ble.sh falls back to the built-in completion generation and uses the man pages to generate options.

> But 1. I want inner command to be shown,

Could you explain in a bit more detail what is *inner command*? If you are talking about subcommands like `vim <inner_command>`, you need a separate completion setting that provides the list of inner commands. I've checked the completion setting for `vim` provided by `bash-completion`, but it doesn't seem to provide completing such *inner commands*. I checked the man page of Vim, but it doesn't seem to describe any command line usage like `vim <inner_command> ...`. I also searched for "*inner command*" in Google, but I don't seem to find it. Some people seem to call the inner text objects as "inner command", but these are the commands used in the text editing in the Vim editor and not supposed to be specified to the command line arguments.

> 2. I don't want command options to be shown without '-'.

This behavior is implemented by `bash-completion`. You can set it up in your `~/.bashrc` (or in `/etc/profile.d` through the system configuration files such as `/etc/bashrc`, etc.).

> I read the discussion here briefly and guess [this commit](https://github.com/akinomyoga/ble.sh/commit/54ace59c57557169ce68366f23849cfe5f8413e8#diff-c180d07f9f6314cd0ce60494fd95e4a1a3afaa2f78afd7046897c4c4e6f072d3R5122-R5141) is about my requirement. But I don't understand how to configure and even cannot sure that code fragment is about this topic.

It is unrelated. That commit introduced a function that users can programmatically get the list of options.

**Reply by moovinzoo** [reply]
Thanks for your replying. And also sorry for my bad example, vim was not correct for the case, as you said it doesn't have any inner_command.
On behalf of your kind explanation now I could understand generating options or commands are about bash-completion's issue, but unfortunetly, I still can't understand then why behavior of completion menu varies time to time. 

As I mensioned early; 
> When I use command vim (tab) or git (tab), it shows options, files for the first time when I enters new shell. But when I delete the space before command and re-insert space and press tab then it shows (sub)commands as expected.

(vim was not proper example so I'll change the command example to git) when I open the bash shell then typped `git <tab>` then completion is not working well by showing mene that has only options and files in current directory. When I remove and re-add preceding space after the command, and then completion now works well with showing commands when pressing (tab), actually at last by the second time.

**Reply by akinomyoga** [reply]
OK, thank you for your clarification!

> but unfortunetly, I still can't understand then why behavior of completion menu varies time to time.

Yeah, that is strange, though I don't seem to be able to reproduce the problem. I've replied to the corresponding thread above: https://github.com/akinomyoga/ble.sh/discussions/436#discussioncomment-9113190.
